src/zkevm_specs/evm/execution/calldataload.py

```diff
@@ -0,0 +1,58 @@
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+from ..table import RW, CallContextFieldTag, TxContextFieldTag
+from ..util import BufferReaderGadget
+from ...util.param import N_BYTES_WORD
+
+
+def calldataload(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+    instruction.constrain_equal(opcode, Opcode.CALLDATALOAD)
+
+    # offset is the 64-bit offset to start reading 32-bytes from start of calldata.
+    offset = instruction.rlc_to_fq_exact(instruction.stack_pop(), n_bytes=8)
+
+    tx_id = instruction.call_context_lookup(CallContextFieldTag.TxId, RW.Read)
+
+    if instruction.curr.is_root:
+        calldata_length = instruction.tx_context_lookup(tx_id, TxContextFieldTag.CallDataLength)
+        calldata_offset = 0
+        src_addr = offset
+        src_addr_end = calldata_length
+    else:
+        calldata_length = instruction.call_context_lookup(CallContextFieldTag.CallDataLength)
+        calldata_offset = instruction.call_context_lookup(CallContextFieldTag.CallDataOffset)
+        src_addr = offset + calldata_offset
+        src_addr_end = calldata_offset + calldata_length
+
+    bytes_left = N_BYTES_WORD if calldata_length.n > src_addr_end.n else src_addr_end - src_addr
```

Comment:
Can we express this in `instruction.select`?
And can we use `src_addr_end` for consistency with the circuit?

Reply:
I tried using `instruction.select` for a few cases like:
```python
calldata_length = instruction.select(
    instruction.curr.is_root,
    instruction.tx_context_lookup(tx_id, TxContextFieldTag.CallDataLength),
    instruction.call_context_lookup(CallContextFieldTag.CallDataLength),
)
```

But this tries to add those two lookups before passing the looked up values to the `select` function.

This leads to lookup failures.

Reply:
I've used the `N_BYTES_WORD` to match the circuits implementation. Thanks for noticing that :)

Reply:
No problem :)
Your way sounds good.